ssl/t1_lib.c

```diff
@@ -3452,3 +3450,27 @@ size_t ssl_hmac_size(const SSL_HMAC *ctx)
     return 0;
 }
 
+#ifndef OPENSSL_NO_EC
+int ssl_get_EC_curve_nid(const EVP_PKEY *pkey)
+{
+    int ret = NID_undef;
+
+    if (EVP_PKEY_id(pkey) != EVP_PKEY_KEYMGMT) {
+        if (EVP_PKEY_base_id(pkey) == EVP_PKEY_EC) {
+            EC_KEY *ec = EVP_PKEY_get0_EC_KEY(pkey);
+
+            ret = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec));
+        }
+    } else {
+        char name[80] = { 0 };
+        size_t name_len;
+
+        if (EVP_PKEY_get_utf8_string_param((EVP_PKEY *)pkey,
```

Comment:
We need to get this to work with legacy keys so that the first "if" branch is not necessary.

Reply:
i.e. by exporting it to a provider if necessary

Reply:
> i.e. by exporting it to a provider if necessary

That's a *lot* of churn just to get the curve name.  While I sympathise with the purity standpoint, there's also something to be said about pragmatism.

Reply:
Something we *could* do is a legacy section in `EVP_PKEY_get_utf8_string_param()` that handles legacy keys...  The inverse of functions like `legacy_ctrl_to_param()` if you will...  *Much* less churn.

Reply:
> Something we could do is a legacy section in EVP_PKEY_get_utf8_string_param() that handles legacy keys... The inverse of functions like legacy_ctrl_to_param() if you will... Much less churn.

That would be ok with me.

Reply:
For another time...